### PR TITLE
Add podman system prune and info commands

### PR DESCRIPTION
**Pull request by rhatdan**
We are missing the equivalence of the docker system commands

This patch set adds `podman system prune`

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

File: cmd/podman/system_prune.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		Description:  pruneSystemDescription,
+		Action:       pruneSystemCmd,
+		OnUsageError: usageErrorHandler,
+		Flags:        pruneSystemFlags,
+	}
+)
+
+func pruneSystemCmd(c *cli.Context) error {
+
+	runtime, err := adapter.GetRuntime(c)
```
Comment (baude) [inline cmd/podman/system_prune.go]: bonus points for using the adapter.Runtime!

**Comment by baude** [comment]
would really like some tests if that is possible

File: cmd/podman/system_prune.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	defer runtime.Shutdown(false)
+	// Call prune; if any cids are returned, print them and then
+	// return err in case an error also came up
+	pruneCids, err := runtime.PruneImages(c.Bool("all"))
+	if len(pruneCids) > 0 {
+		for _, cid := range pruneCids {
+			fmt.Println(cid)
+		}
+	}
+	err1 := volumePrune(runtime, getContext(), c.Bool("force"))
```
Comment (mheon) [inline cmd/podman/system_prune.go]: I don't think `docker system prune` actually prunes volumes. The description says exclusively containers, images, networks

Comment (vrothberg) [reply cmd/podman/system_prune.go]: It does, see https://docs.docker.com/engine/reference/commandline/system_prune/

Comment (vrothberg) [reply cmd/podman/system_prune.go]: There's a `--volume` flag.

**Comment by rhatdan** [comment]
@vrothberg @mheon @baude @umohnani8 @giuseppe @TomSweeneyRedHat PTAL
@baude Now has tests.

File: cmd/podman/system_prune.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	ctx := getContext()
+	fmt.Println("Deleted Containers")
+	err = pruneContainers(runtime, ctx, shared.Parallelize("rm"), c.Bool("force"))
+	if c.Bool("volumes") {
+		fmt.Println("Deleted Volumes")
+		err1 := volumePrune(runtime, getContext())
```
Comment (mheon) [inline cmd/podman/system_prune.go]: Should we be printing IDs here?

Comment (rhatdan) [reply cmd/podman/system_prune.go]: volumePrune should print the ids that it is removing.

**Comment by mheon** [comment]
LGTM

File: cmd/podman/system_prune.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+var (
+	pruneSystemDescription = `
+	podman system prune
+
+        Remove unused data
+`
+	pruneSystemFlags = []cli.Flag{
+		cli.BoolFlag{
+			Name:  "all, a",
+			Usage: "remove all unused data",
```
Comment (vrothberg) [inline cmd/podman/system_prune.go]: s/remove/Remove/ for consistency

File: cmd/podman/system_prune.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+			Usage: "Do not prompt for confirmation",
+		},
+		cli.BoolFlag{
+			Name:  "volumes",
+			Usage: "Prune volumes",
+		},
+	}
+	pruneSystemCommand = cli.Command{
+		Name:         "prune",
+		Usage:        "Remove unused data",
```
Comment (vrothberg) [inline cmd/podman/system_prune.go]: "data" versus "systems"?

File: cmd/podman/system_prune.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		volumeString := ""
+		if c.Bool("volumes") {
+			volumeString = `
+        - all volumes not used by at least one container`
+		}
+		fmt.Printf(`
+WARNING! This will remove:
+        - all stopped containers%s
+        - all dangling images
+        - all build cache
```
Comment (vrothberg) [inline cmd/podman/system_prune.go]: Can we add `- all volumes` if the flag is set?

File: cmd/podman/system_prune.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+	runtime, err := adapter.GetRuntime(c)
+	if err != nil {
+		return errors.Wrapf(err, "could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	ctx := getContext()
+	fmt.Println("Deleted Containers")
+	err = pruneContainers(runtime, ctx, shared.Parallelize("rm"), c.Bool("force"))
```
Comment (vrothberg) [inline cmd/podman/system_prune.go]: s/err/errPruneContainers/

File: cmd/podman/system_prune.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+				logrus.Errorf("%q", err1)
+			} else {
+				err = err1
+			}
+		}
+	}
+
+	// Call prune; if any cids are returned, print them and then
+	// return err in case an error also came up
+	pruneCids, err1 := runtime.PruneImages(c.Bool("all"))
```
Comment (vrothberg) [inline cmd/podman/system_prune.go]: s/err1/err/PruneImages/ ... just to make the code easier to brain-parse

File: docs/podman-system-prune.1.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+% PODMAN(1) Podman Man Pages
+% Brent Baude
+% December 2018
```
Comment (vrothberg) [inline docs/podman-system-prune.1.md]: Copy artifact?

File: docs/podman-system-prune.1.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+[**-help**|**--h**]
+[**-volumes**|**--v**]
+
+# DESCRIPTION
+**podman system prune** removes all dangling systems from local storage. With the `all` option,
+you can delete all unused systems.  Unused systems are dangling systems as well as any system that
+does not have any containers based on it.
+
+By default, volumes are not removed to prevent important data from being deleted if there is currently no container using the volume. Use the --volumes flag when running the command to prune volumes as well.
+
```
Comment (vrothberg) [inline docs/podman-system-prune.1.md]: Can we define what we mean by **systems** (i.e., containers, images, build cache)? Same applies to the help message. Currently, it's hard to figure out which data is removed without actually running the command.

File: cmd/podman/system_prune.go (hunk 7 above)
Comment (vrothberg) [reply cmd/podman/system_prune.go]: I am not sure if we should use `--force` here as it should only impact the cli prompt.

File: cmd/podman/system_prune.go (hunk 8 above)
Comment (rhatdan) [reply cmd/podman/system_prune.go]: Ok I changed to use lasterr to make it clearer.

File: docs/podman-system-prune.1.md (hunk 9 above)
Comment (rhatdan) [reply docs/podman-system-prune.1.md]: Fixed

File: cmd/podman/system_prune.go (hunk 4 above)
Comment (rhatdan) [reply cmd/podman/system_prune.go]: Fixed these all in the second patch.

File: cmd/podman/system_prune.go (hunk 5 above)
Comment (rhatdan) [reply cmd/podman/system_prune.go]: Data is correct.  systems was incorrect.

File: cmd/podman/system_prune.go (hunk 7 above)
Comment (rhatdan) [reply cmd/podman/system_prune.go]: Removed

**Comment by giuseppe** [comment]
LGTM

File: cmd/podman/pod_unpause.go [hunk 11]
```diff
@@ -13,7 +13,7 @@ var (
 	podUnpauseFlags = []cli.Flag{
 		cli.BoolFlag{
 			Name:  "all, a",
-			Usage: "unpause all paused pods",
+			Usage: "Anpause all paused pods",
```
Comment (TomSweeneyRedHat) [inline cmd/podman/pod_unpause.go]: s/Anpause/Unpause/

File: docs/podman-info.1.md [hunk 12]
```diff
@@ -1,9 +1,9 @@
 % podman-version(1)
```
Comment (TomSweeneyRedHat) [inline docs/podman-info.1.md]: Should this be podman-info?

File: docs/podman-info.1.md [hunk 13]
```diff
@@ -1,9 +1,9 @@
...
 
 ## NAME
+podman\-system\-info - Display system information
```
Comment (TomSweeneyRedHat) [inline docs/podman-info.1.md]: Should this have it's own page, even if it's mostly a cut/paste job?

Comment (rhatdan) [reply docs/podman-info.1.md]: No they are identical,  This is what we do with all of the podman container commands

podman-container-ps -> podman-ps.
for example.  I think we should add the alternate names to the linked images though.

File: docs/podman-system.1.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+**podman system** *subcommand*
+
+## DESCRIPTION
+The system command allows you to manage the podman systems
+
+## COMMANDS
+
+| Command  | Man Page                                            | Description                                                                  |
+| -------  | --------------------------------------------------- | ---------------------------------------------------------------------------- |
+| info     | [podman-system-info(1)](podman-system-info.1.md)    | Displays Podman related system information.                                  |
```
[the remark this answers is not in the document]
Comment (rhatdan) [reply docs/podman-system.1.md]: Fixed

File: docs/podman-info.1.md (hunk 12 above)
Comment (rhatdan) [reply docs/podman-info.1.md]: Yup, Fixed

File: cmd/podman/pod_unpause.go (hunk 11 above)
Comment (rhatdan) [reply cmd/podman/pod_unpause.go]: Fixed

**Pull request merged**